### PR TITLE
feat(cb2-4202): amend test result for specialist tests

### PR DESCRIPTION
**Pull request by tahayazici**
## Amend Test Result (Specialist Tests)

- Update the dynamic forms service to support async validators.
- Validate the result of specialist tests against their custom defects.

Big thanks to @mcadenas-bjss for cracking the problem of handling async validators.

[CB2-4202](https://dvsa.atlassian.net/browse/CB2-4202)

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

**Review by gjulien-bjss — COMMENTED** [review]
All good apart from failing tests

File: src/app/forms/templates/general/defect.template.ts [hunk 1]
```diff
@@ -136,9 +136,9 @@ export const DefectsTpl: FormNode = {
                     {
                       name: 'prs',
                       label: 'PRS',
+                      value: null,
                       type: FormNodeTypes.CONTROL,
                       editType: FormNodeEditTypes.RADIO,
-                      value: null,
```
Comment (gjulien-bjss) [inline src/app/forms/templates/general/defect.template.ts:141]: These values will need to be initialised as `null` for add defect to work as intended

Comment (tahayazici) [reply src/app/forms/templates/general/defect.template.ts:141]: Oh, I did not remove it. Just that in my branch `value: null,` was a few lines above compared to yours.

Comment (gjulien-bjss) [reply src/app/forms/templates/general/defect.template.ts:141]: 🤦

File: src/app/forms/validators/custom-validators.ts [hunk 2]
```diff
@@ -133,4 +136,25 @@ export class CustomValidators {
   static invalidOption: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
     return '[INVALID_OPTION]' === control.value ? { invalidOption: true } : null;
   };
+
+  static resultDependantOnCustomDefects(store: Store<TestResultsState>): AsyncValidatorFn {
```
Comment (gjulien-bjss) [inline src/app/forms/validators/custom-validators.ts:140]: Also, there is no unit test for the validator you've added. The async validators could also be put in a separate file/class if they have a different enum and map, might as well.

File: src/app/forms/validators/custom-async-validators.spec.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    const result = await firstEmitted(
+      CustomAsyncValidators.resultDependantOnCustomDefects(store)(form.controls['testResult']) as Observable<ValidationErrors | null>
+    );
+
+    expect(result).toBeNull();
+  });
+});
+
+function firstEmitted<T>(obs$: Observable<T>): Promise<T> {
```
Comment (gjulien-bjss) [inline src/app/forms/validators/custom-async-validators.spec.ts:82]: Can you not use the `firstValueFrom` operator ?

File: src/app/forms/validators/custom-validators.ts (hunk 2 above)
Comment (tahayazici) [reply src/app/forms/validators/custom-validators.ts]: All done, nice suggestion.

Comment (tahayazici) [reply src/app/forms/validators/custom-validators.ts]: All done, nice suggestion.

**Review by mcadenas-bjss — COMMENTED** [review]
will approve after change using `firstValueFrom`

File: src/app/forms/validators/custom-async-validators.spec.ts (hunk 3 above)
Comment (tahayazici) [reply src/app/forms/validators/custom-async-validators.spec.ts]: This is great - suggestion applied.

**Comment by tahayazici** [comment]
Now using `firstValueFrom`, @mcadenas-bjss.

**Review by gjulien-bjss — DISMISSED** [review]

**Review by mcadenas-bjss — APPROVED** [review]

**Review by gjulien-bjss — APPROVED** [review]

**Review by Ruairidh-BJSS — APPROVED** [review]

**Review by mcadenas-bjss — APPROVED** [review]

**Pull request merged**